Libplanet/Net/BlockCompletion.cs

```diff
@@ -0,0 +1,466 @@
+                    "(Remained incomplete demands: {IncompleteDemands})",
+                    block.Index,
+                    block.Hash,
+                    _demands.Count + _satisfiedBlocks.Count(kv => !kv.Value)
+                );
+                return true;
+            }
+
+            if (_satisfiedBlocks.ContainsKey(block.Hash))
+            {
+                _logger.Verbose(
+                    "The block #{BlockIndex} {BlockHash} is already complete. " +
+                    "(Remained incomplete demands: {IncompleteDemands})",
+                    block.Index,
+                    block.Hash,
+                    _demands.Count + _satisfiedBlocks.Count(kv => !kv.Value)
+                );
+                return false;
+            }
+
+            _logger.Verbose(
+                "The block #{BlockIndex} {BlockHash} has never demanded. " +
+                "(Remained incomplete demands: {IncompleteDemands})",
+                block.Index,
+                block.Hash,
+                _demands.Count + _satisfiedBlocks.Count(kv => !kv.Value)
+            );
+            return false;
+        }
+
+        /// <summary>
+        /// Downloads blocks from <paramref name="peers"/> in parallel,
+        /// using the given <paramref name="blockFetcher"/> function.
+        /// </summary>
+        /// <param name="peers">A list of peers to download blocks.</param>
+        /// <param name="blockFetcher">A function to take demands and a peer, and then
+        /// download corresponding blocks.</param>
+        /// <param name="singleSessionTimeout">A maximum time to wait each single call of
+        /// <paramref name="blockFetcher"/>.  If a call is timed out unsatisfied demands
+        /// are automatically retried to fetch from other peers.</param>
+        /// <param name="cancellationToken">A cancellation token to observe while waiting
+        /// for the task to complete.</param>
+        /// <returns>An async enumerable that yields pairs of a fetched block and its source
+        /// peer.  It terminates when all demands are satisfied.</returns>
+        public async IAsyncEnumerable<Tuple<Block<TAction>, TPeer>> Complete(
+            IReadOnlyList<TPeer> peers,
+            BlockFetcher blockFetcher,
+            TimeSpan singleSessionTimeout,
+            [EnumeratorCancellation] CancellationToken cancellationToken = default
+        )
+        {
+            if (!peers.Any())
+            {
+                throw new ArgumentException("The list of peers must not be empty.", nameof(peers));
+            }
+
+            PeerPool pool = new PeerPool(peers);
+            var queue = new AsyncProducerConsumerQueue<Tuple<Block<TAction>, TPeer>>();
+            var completion =
+                new ConcurrentDictionary<HashDigest<SHA256>, bool>(_satisfiedBlocks);
+
+            await foreach (var hashes in EnumerateChunks(cancellationToken))
+            {
+                foreach (HashDigest<SHA256> hash in hashes)
```

Comment:
Shouldn't [this](https://github.com/planetarium/libplanet/pull/798/files#diff-a155979ce9e3dfbb6ba0ea031a5b7a44R214) be done before here and use the `hashDigests` instead of `hashs`?